File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/custom/VisionEncodePatches.java
/*
 *  ******************************************************************************
 *  *
 *  *
 *  * This program and the accompanying materials are made available under the
 *  * terms of the Apache License, Version 2.0 which is available at
 *  * https://www.apache.org/licenses/LICENSE-2.0.
 *  *
 *  *  See the NOTICE file distributed with this work for additional
 *  *  information regarding copyright ownership.
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 *  * License for the specific language governing permissions and limitations
 *  * under the License.
 *  *
 *  * SPDX-License-Identifier: Apache-2.0
 *  *****************************************************************************
 */

package org.nd4j.linalg.api.ops.impl.transforms.custom;

import lombok.NoArgsConstructor;
import org.nd4j.autodiff.samediff.SDVariable;
import org.nd4j.autodiff.samediff.SameDiff;
import org.nd4j.common.base.Preconditions;
import org.nd4j.linalg.api.buffer.DataType;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.api.ops.DynamicCustomOp;

import java.util.Arrays;
import java.util.List;

/**
 * ViT-style patch extraction for Vision-Language Model training.
 *
 * <p>Splits an image into non-overlapping square patches and flattens each patch
 * into a 1-D feature vector. This is the first step of vision encoding used in
 * VLMs such as LLaVA, Idefics, and SmolDocling (ViT-style backbones).</p>
 *
 * <p>Given an image of shape {@code [batch, channels, height, width]} and a
 * {@code patchSize}, the output is:</p>
 * <ul>
 *   <li>Output 0: {@code [batch, numPatches, patchDim]} where
 *       {@code numPatches = (height/patchSize) * (width/patchSize)} and
 *       {@code patchDim = channels * patchSize * patchSize}</li>
 *   <li>Output 1: INT64 scalar — {@code numPatches} per image</li>
 * </ul>
 *
 * <p>The image height and width must both be divisible by {@code patchSize}.</p>
 *
 * <p>Example:</p>
 * <pre>{@code
 * // [1, 3, 224, 224] image with patchSize=14 → [1, 256, 588] patches
 * // numPatches = (224/14)^2 = 256, patchDim = 3*14*14 = 588
 * INDArray image = Nd4j.rand(DataType.FLOAT, 1, 3, 224, 224);
 * INDArray[] out = Nd4j.exec(new VisionEncodePatches(image, 14));
 * INDArray patches = out[0];  // shape [1, 256, 588]
 * INDArray nPatches = out[1]; // scalar 256
 *
 * // In SameDiff:
 * SameDiff sd = SameDiff.create();
 * SDVariable img = sd.placeHolder("image", DataType.FLOAT, 1, 3, 224, 224);
 * SDVariable[] out = new VisionEncodePatches(sd, img, 14).outputVariables();
 * SDVariable patches = out[0];
 * SDVariable numPatches = out[1];
 * }</pre>
 *
 * @author Adam Gibson
 */
@NoArgsConstructor
public class VisionEncodePatches extends DynamicCustomOp {

    private int patchSize = 14;
    private int hiddenDim = 0;

    /**
     * Construct a patch-extraction op.
     *
     * @param image     input image {@code [batch, channels, height, width]}
     * @param patchSize size of each square patch (e.g. 14); height and width must
     *                  each be divisible by this value
     */
    public VisionEncodePatches(INDArray image, int patchSize) {
        this(image, patchSize, 0);
    }

    /**
     * Construct a patch-extraction op with an explicit hidden dimension hint.
     *
     * @param image     input image {@code [batch, channels, height, width]}
     * @param patchSize size of each square patch (e.g. 14)
     * @param hiddenDim reserved for downstream projection ops (pass 0 if unused)
     */
    public VisionEncodePatches(INDArray image, int patchSize, int hiddenDim) {
        super(wrapFilterNull(image), null);
        this.patchSize = patchSize;
        this.hiddenDim = hiddenDim;
        addIArgument(patchSize, hiddenDim);
    }

    /**
     * Construct a SameDiff patch-extraction op.
     *
     * @param sd        the SameDiff instance
     * @param image     image variable {@code [batch, channels, height, width]}
     * @param patchSize size of each square patch (e.g. 14)
     */
    public VisionEncodePatches(SameDiff sd, SDVariable image, int patchSize) {
        this(sd, image, patchSize, 0);
    }

    /**
     * Construct a SameDiff patch-extraction op with a hidden dimension hint.
     *
     * @param sd        the SameDiff instance
     * @param image     image variable {@code [batch, channels, height, width]}
     * @param patchSize size of each square patch (e.g. 14)
     * @param hiddenDim reserved for downstream projection ops (pass 0 if unused)
     */
    public VisionEncodePatches(SameDiff sd, SDVariable image, int patchSize, int hiddenDim) {
        super(null, sd, new SDVariable[]{image}, false);
        this.patchSize = patchSize;
        this.hiddenDim = hiddenDim;
        addIArgument(patchSize, hiddenDim);
    }

    @Override
    public void configureFromArguments() {
        super.configureFromArguments();
        if (!iArguments.isEmpty()) {
            this.patchSize = iArguments.get(0).intValue();
        }
        if (iArguments.size() > 1) {
            this.hiddenDim = iArguments.get(1).intValue();
        }
    }

    @Override
    public String opName() {
        return "vision_encode_patches";
    }

    /**
     * Returns the patch size used by this op.
     *
     * @return patch size
     */
    public int getPatchSize() {
        return patchSize;
    }

    /**
     * Returns the hidden dim hint (may be 0 if unused).
     *
     * @return hidden dim
     */
    public int getHiddenDim() {
        return hiddenDim;
    }

    @Override
    public List<DataType> calculateOutputDataTypes(List<DataType> dataTypes) {
        Preconditions.checkState(!dataTypes.isEmpty(),
                "Expected at least 1 input data type, got: %s", dataTypes);
        DataType inputType = dataTypes.get(0);
        Preconditions.checkState(inputType.isFPType(),
                "Input data type must be a floating point type for vision_encode_patches, got: %s",
                inputType);
        return Arrays.asList(inputType, DataType.INT64);
    }

    @Override
    public int getNumOutputs() {
        return 2;
    }
}

File: libnd4j/include/ops/declarable/generic/nn/vision_encode_patches.cpp
/* ******************************************************************************
 *
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 *  See the NOTICE file distributed with this work for additional
 *  information regarding copyright ownership.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

//
// @author Adam Gibson
//
// vision_encode_patches: ViT-style patch extraction from images.
// Splits an image into non-overlapping patches and flattens each patch
// into a 1-D feature vector. This is the first step of ViT-style vision
// encoding used in VLMs such as LLaVA, Idefics, and SmolDocling.
//

#include <system/op_boilerplate.h>
#if NOT_EXCLUDED(OP_vision_encode_patches)

#include <ops/declarable/headers/vlm.h>
#include <helpers/ShapeUtils.h>

namespace sd {
namespace ops {

//////////////////////////////////////////////////////////////////////////
// vision_encode_patches
//
// Input:
//   0: image  [batch, channels, height, width]  — input image(s)
// Int args:
//   0: patch_size  — size of each square patch (e.g. 14)
//   1: hidden_dim  — unused in the extract step; kept for API symmetry with
//                    downstream projection ops (default: 0 = patch_size^2 * channels)
// Outputs:
//   0: patches [batch, num_patches, patch_size*patch_size*channels]
//   1: num_patches_per_image  — INT64 scalar (= (H/patch_size)*(W/patch_size))
//////////////////////////////////////////////////////////////////////////
template <typename T>
static Status vision_encode_patches_impl(NDArray* image, NDArray* patches, NDArray* numPatchesOut,
                                          int patchSize) {
    const LongType batch    = image->sizeAt(0);
    const LongType channels = image->sizeAt(1);
    const LongType height   = image->sizeAt(2);
    const LongType width    = image->sizeAt(3);

    const LongType numPatchesH = height / patchSize;
    const LongType numPatchesW = width  / patchSize;
    const LongType numPatches  = numPatchesH * numPatchesW;
    const LongType patchDim    = channels * patchSize * patchSize;

    // Write the scalar num_patches output
    double numPatchesVal = static_cast<double>(numPatches);
    numPatchesOut->assign(numPatchesVal);

    auto imageBuf   = image->bufferAsT<T>();
    auto patchesBuf = patches->bufferAsT<T>();

    PRAGMA_OMP_PARALLEL_FOR
    for (LongType b = 0; b < batch; ++b) {
        for (LongType ph = 0; ph < numPatchesH; ++ph) {
            for (LongType pw = 0; pw < numPatchesW; ++pw) {
                const LongType patchIdx   = ph * numPatchesW + pw;
                const LongType outOffset  = (b * numPatches + patchIdx) * patchDim;

                for (LongType c = 0; c < channels; ++c) {
                    for (LongType i = 0; i < patchSize; ++i) {
                        const LongType inRow    = ph * patchSize + i;
                        const LongType inBase   = ((b * channels + c) * height + inRow) * width
                                                   + pw * patchSize;
                        const LongType patchBase = (c * patchSize + i) * patchSize;
                        PRAGMA_OMP_SIMD
                        for (LongType j = 0; j < patchSize; ++j) {
                            patchesBuf[outOffset + patchBase + j] = imageBuf[inBase + j];
                        }
                    }
                }
            }
        }
    }

    return Status::OK;
}

CUSTOM_OP_IMPL(vision_encode_patches, 1, 2, false, 0, 2) {
    auto image = INPUT_VARIABLE(0);
    auto patches = OUTPUT_VARIABLE(0);
    auto numPatchesOut = OUTPUT_VARIABLE(1);

    REQUIRE_TRUE(image->rankOf() == 4, 0,
        "vision_encode_patches: expected 4-D image [batch, channels, H, W], got rank %d",
        image->rankOf());

    int patchSize = INT_ARG(0);
    // INT_ARG(1) is hidden_dim — not used in patch extraction

    REQUIRE_TRUE(patchSize > 0, 0,
        "vision_encode_patches: patch_size must be positive, got %d", patchSize);

    const LongType height = image->sizeAt(2);
    const LongType width  = image->sizeAt(3);

    REQUIRE_TRUE(height % patchSize == 0 && width % patchSize == 0, 0,
        "vision_encode_patches: image dimensions (%lld x %lld) must be divisible by "
        "patch_size (%d)", (long long)height, (long long)width, patchSize);

    auto imageType = image->dataType();
    BUILD_SINGLE_SELECTOR(imageType, return vision_encode_patches_impl,
                          (image, patches, numPatchesOut, patchSize), SD_FLOAT_TYPES);
}

DECLARE_SHAPE_FN(vision_encode_patches) {
    auto inShape  = inputShape->at(0);
    auto dtype    = ArrayOptions::dataType(inShape);

    int patchSize = INT_ARG(0);

    const LongType batch    = shape::sizeAt(inShape, static_cast<LongType>(0));
    const LongType channels = shape::sizeAt(inShape, static_cast<LongType>(1));
    const LongType height   = shape::sizeAt(inShape, static_cast<LongType>(2));
    const LongType width    = shape::sizeAt(inShape, static_cast<LongType>(3));

    const LongType numPatches = (height / patchSize) * (width / patchSize);
    const LongType patchDim   = channels * patchSize * patchSize;

    // Output 0: [batch, num_patches, patch_dim]
    auto patchShape = ConstantShapeHelper::getInstance().createShapeInfo(
        dtype, 'c', {batch, numPatches, patchDim});

    // Output 1: scalar INT64
    auto numPatchesShape = ConstantShapeHelper::getInstance().scalarShapeInfo(DataType::INT64);

    return SHAPELIST(patchShape, numPatchesShape);
}

DECLARE_TYPES(vision_encode_patches) {
    getOpDescriptor()->addTraits(OP_TRAIT_FULLY_WRITING);
    getOpDescriptor()
        ->setAllowedInputTypes({ALL_FLOATS})
        ->setAllowedOutputTypes(0, {ALL_FLOATS})
        ->setAllowedOutputTypes(1, {DataType::INT64});
}

}  // namespace ops
}  // namespace sd

#endif  // NOT_EXCLUDED(OP_vision_encode_patches)
